codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/AssembleMiddlewareStack.java

```diff
@@ -83,78 +84,63 @@ private void writeAddMiddlewareBefore(
+        // Option symbol
+        Symbol optionSymbol = SymbolUtils.createValueSymbolBuilder("options").build();
         return ListUtils.of(
                 // Add RequestInvocationIDMiddleware to operation stack
                 RuntimeClientPlugin.builder()
-                    .buildMiddlewareStack((writer, service, operation, protocolGenerator, stackOperand) -> {
-                        // RequestInvocationIDMiddleware
-                        Symbol RequestInvocationIDMiddleware = SymbolUtils.createValueSymbolBuilder(
-                                "RequestInvocationIDMiddleware", AwsGoDependency.AWS_MIDDLEWARE
-                        ).build();
-
-                        writeAddMiddlewareAfter(writer, BUILD_MIDDLEWARE,
-                                RequestInvocationIDMiddleware, "$T{}", stackOperand);
-                    }).build(),
-
-                // Add serializer middleware to operation stack
+                        .registerMiddleware(MiddlewareRegistrar.builder()
+                                .resolvedFunction(SymbolUtils.createValueSymbolBuilder(
+                                        "AddRequestInvocationIDMiddleware", AwsGoDependency.AWS_MIDDLEWARE).build())
+                                .build()
+                        )
+                        .build(),
+
+                // Add protocol middleware to operation stack
                 RuntimeClientPlugin.builder()
-                        .buildMiddlewareStack((writer, service, operation, protocolGenerator, stackOperand) -> {
-                            if (protocolGenerator == null){
-                                return;
-                            }
-                            String serializerMiddlewareName = ProtocolGenerator.getSerializeMiddlewareName(operation.getId(),
-                                    protocolGenerator.getProtocolName());
-                            writeAddMiddlewareAfter(writer, SERIALIZE_MIDDLEWARE,
-                                    serializerMiddlewareName, "&$L{}", stackOperand);
-                        }).build(),
-
-                // Add deserializer middleware to operation stack
-                RuntimeClientPlugin.builder()
-                        .buildMiddlewareStack((writer, service, operation, protocolGenerator, stackOperand) -> {
-                            if (protocolGenerator == null) {
-                                return;
-                            }
-                            String deserializerMiddlewareName = ProtocolGenerator.getDeserializeMiddlewareName(operation.getId(),
-                                    protocolGenerator.getProtocolName());
-                            writeAddMiddlewareAfter(writer, DESERIALIZE_MIDDLEWARE,
-                                    deserializerMiddlewareName, "&$L{}", stackOperand);
-                        }).build(),
+                        .writeMiddlewareHelpers((writer, service, operation, protocolGenerator) -> {
```

Comment:
I didn't catch this on the first round, but we should be adding the serde functions to all operations across protocols, so this should be generated in `OperationGenerator`.